Mu2eG4/src/GammaConvGenerator_module.cc

```diff
@@ -0,0 +1,214 @@
+      fhicl::Atom<std::string> stopMaterial{Name("stopMaterial"),Comment("G4 name of the material where the photon has converted (temporary)."),"ST_Wires"};
+      fhicl::Atom<unsigned> verbosity{Name("verbosity")};
+      fhicl::Atom<bool>   useCorrelatedAngleOverKE{Name("useCorrelatedAngleOverKE"), Comment("Flag to use correlated e+e- cos/KE"), false};
+    };
+
+    using Parameters= art::EDProducer::Table<Config>;
+    explicit GammaConvGenerator(const Parameters& conf);
+
+    virtual void produce(art::Event& event) override;
+
+    //----------------------------------------------------------------
+  private:
+
+    double particleMass_;
+
+    // Module label of the module that made the volume.
+    std::string physVolModuleLabel_;
+    // Module label of the last process that wrote the volume list.
+    std::string physVolProcessLabel_;
+
+    art::ProductToken<SimParticleCollection> const simsToken_;
+    std::string stopMaterial_;
+
+    unsigned verbosity_;
+
+    art::RandomNumberGenerator::base_engine_t& eng_;
+    CLHEP::RandFlat randFlat_;
+    bool   useCorrelatedAngleOverKE_; //correlate the cos/ke for the e+ and e-
+
+    GammaPairConversionSpectrum* spectrum_; //pair production spectrum
+    enum {kMaxConversionMaterialElements = 10};
+    ProcessCode process;
+  };
+
+  //================================================================
+  GammaConvGenerator::GammaConvGenerator(const Parameters& conf)
+    : EDProducer{conf}
+      //    , particleMass_(GlobalConstantsHandle<ParticleDataList>()->particle(static_cast<PDGCode::type>(PDGCode::e_minus)).mass())
+    , physVolModuleLabel_{conf().physVolModuleLabel()}
+    , physVolProcessLabel_{conf().physVolProcessLabel()}
+    , simsToken_{consumes<SimParticleCollection>(conf().inputSimParticles())}
+    , stopMaterial_{conf().stopMaterial()}
+    , verbosity_{conf().verbosity()}
+    , eng_{createEngine(art::ServiceHandle<SeedService>()->getSeed())}
+    , randFlat_{eng_}
+    , useCorrelatedAngleOverKE_(conf().useCorrelatedAngleOverKE())
+    , spectrum_                (new GammaPairConversionSpectrum(&randFlat_, useCorrelatedAngleOverKE_))
+  {
+    produces<mu2e::StageParticleCollection>();
+    process = ProcessCode::mu2eGammaConversion;
+  }
+
+  //================================================================
+  void GammaConvGenerator::produce(art::Event& event) {
+    auto output{std::make_unique<StageParticleCollection>()};
+
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_);
+    const auto gammas = simParticleList(simh,PDGCode::gamma,ProcessCode::conv);
+
+    if(gammas.empty()) {
+      throw   cet::exception("NO GAMMAS")
+        <<"GammaConvGenerator::produce(): no suitable converted gamma in the input SimParticleCollection\n";
+    }
+
+    // Take the converted gamma as SimParticle
+    const auto gammapart = gammas.at(eng_.operator unsigned int() % gammas.size());
```

Comment:
This is not guaranteed to uniformly sample the set of gammas.   The effect is negligible.  Did you do it this way since it's advertised as being faster than randFlat_,fireInt( gammas.size());

Reply:
I have inherited this from Michael's code. If you think it's better I can use randFlat.

Reply:
changing to what is suggested